### PR TITLE
varlink: add glue to allow authenticating varlink connections via polkit

### PR DESCRIPTION
**Pull request by bluca**
The missing bits from https://github.com/systemd/systemd/pull/30189 with pidfd+uid instead of pid+starttime+uid, an actual action config file for sd-creds, a fix for polkit's uid dbus type and a fix to avoid leaking creds as journal logs

### CODE REVIEW

File: src/shared/varlink.c [hunk 1]
```diff
@@ -1781,7 +1779,10 @@ static int varlink_format_json(Varlink *v, JsonVariant *m) {
         if (v->output_buffer_size + r + 1 > VARLINK_BUFFER_MAX)
                 return -ENOBUFS;
 
-        varlink_log(v, "Sending message: %s", text);
+        p = json_variant_by_key(m, "parameters");
+        sensitive = json_variant_is_sensitive(m) || json_variant_is_sensitive(ASSERT_PTR(p));
```
Comment (poettering) [inline src/shared/varlink.c]: hmm, it's not sufficient to check these two fields. we must go recursively down the tree.

maybe add a new flag JSON_FORMAT_REFUSE_SENSITIVE which makes json_variant_format() fail with a recognizable error if there's any sensitive field in the whole object tree. Then we'd call it twice: first with the flag. If it works, great we use it and know it's not sensitive. If it fails we call it again without the flag, and know it is sensitive.

By doing this via json_variant_format() we don#t need to scan the whole tree twice in the usual case, only if the thing is marked sensitive, which hopefully is the exception, not the common case.

Comment (bluca) [reply src/shared/varlink.c]: ok, implemented as requested

File: src/shared/json.c [hunk 2]
```diff
@@ -1786,6 +1786,11 @@ int json_variant_format(JsonVariant *v, JsonFormatFlags flags, char **ret) {
         if (flags & JSON_FORMAT_OFF)
                 return -ENOEXEC;
 
+        if ((flags & JSON_FORMAT_REFUSE_SENSITIVE))
+                for (size_t i = 0; i < json_variant_elements(v); ++i)
+                        if (json_variant_is_sensitive(json_variant_by_index(v, i)))
+                                return -EPERM;
```
Comment (poettering) [inline src/shared/json.c]: this isn't a recursive check. it only checks the first level down, not even the top-level...

i.e. maybe add a new helper json_variant_contains_sensitive() which goes down the tree.

(note that JsonVariants are generally immutable for their content, but not their metadata, i.e. we allow the ref counter to change, and we allow the sensitive to be turned on (but not off) any time, and that means we unfortunately really have to do this check recursively...)

**Comment by poettering** [comment]
I think the latter two commits would better be excluded from this PR and be merged in another PR, since we can merge them right-away.

File: src/shared/json.c (hunk 2 above)
Comment (bluca) [reply src/shared/json.c]: Yeah misunderstood how this stuff was stored, fixed now

File: src/shared/json.c [hunk 3]
```diff
@@ -1771,6 +1771,28 @@ static int json_format(FILE *f, JsonVariant *v, JsonFormatFlags flags, const cha
...
 }
 
+static bool json_variant_is_sensitive_recursive(JsonVariant *v) {
+        if (!v)
+                return false;
+        if (json_variant_is_sensitive(v))
+                return true;
+        if (v == JSON_VARIANT_MAGIC_EMPTY_ARRAY ||
+            v == JSON_VARIANT_MAGIC_EMPTY_OBJECT)
+                return false;
```
Comment (poettering) [inline src/shared/json.c:1781]: this check is unnecessary, the immediately following check checks exactly this already.

**Pull request merged**